# COGS 108 - Data Checkpoint

## Authors

- Jinhui Qiu: Conceptualization, Background research, Writing - original draft, Writing - review & editing
- Yifeng Zheng: Data curation, Methodology, Software
- Pingzhang Xu: Analysis, Visualization, Writing - original draft
- Xingchun Wang: Project administration, Background research, Writing - review & editing

## Research Question

Our project asks: To what extent can Spotify audio features predict or explain a track's Spotify popularity score? We will use Spotify popularity score as the main outcome variable, and we will use audio features such as danceability, energy, valence, tempo, acousticness, speechiness, instrumentalness, liveness, loudness, and duration as the main predictor variables. Through regression analysis, we want to understand whether these audio features have a clear relationship with popularity score and which features may be more strongly associated with popularity.

In addition to the main popularity prediction question, we will also use the Most Streamed Spotify Songs 2024 dataset as a comparison dataset. This dataset helps us identify recent hit songs. We will try to match songs from the 2024 dataset to the Spotify Tracks Dataset using track name and artist. If the matching process works well enough, we will compare the audio features of these 2024 hit songs with songs from the broader Spotify track sample.

In this project, genre is not our main research target, and we are not trying to predict a song's genre. Genre will only be used as a background variable to help us understand the dataset, check possible genre imbalance, and explain limitations in the discussion. Our main question is still about the relationship between Spotify audio features and popularity score, and whether hit songs show different audio-feature patterns from the broader track sample.

## Background and Prior Work

Spotify is one of the major music streaming platforms, so Spotify data can be useful for studying the relationship between song features and popularity. In this project, we focus on Spotify popularity score. Spotify's Web API describes track popularity as a value from 0 to 100, with 100 being the most popular. The score is based mostly on the total number of plays a track has and how recent those plays are. Therefore, popularity score is not exactly the same as total lifetime streams, but it is still a useful outcome variable for comparing tracks on Spotify.

Prior research suggests that audio features can help explain part of song popularity. For example, Saragih's study, "Predicting Song Popularity Based on Spotify's Audio Features," uses Spotify audio features and machine learning methods to study song popularity prediction. This supports our choice to use variables such as danceability, energy, loudness, tempo, valence, and acousticness as predictors. These variables describe parts of a song's rhythm, energy, volume, mood, and sound style, so they may be related to how often listeners play a song.

However, predicting hit songs is not a simple problem. Pachet and Roy's work, "Hit Song Science Is Not Yet a Science," argues that predicting song popularity from acoustic or human-labeled features is difficult. This point is important for our project because a song's popularity is not only determined by the sound of the music itself. Other factors, such as artist fame, playlist placement, marketing, release timing, TikTok trends, and platform recommendation systems, can also affect popularity. Many of these factors are not fully captured by the audio features in the Spotify Tracks Dataset.

Other research also shows that Spotify audio features are useful for describing music. Duman et al. discuss Spotify features such as danceability, energy, valence, tempo, loudness, speechiness, and acousticness as measurable ways to describe musical properties and listening experiences. This supports our use of Spotify audio features as meaningful variables, but it does not mean these features directly cause popularity.

The research gap our project addresses is how much Spotify audio features can actually explain popularity score. These features may provide useful prediction information, but they may only explain part of why a song becomes popular. We also want to know whether 2024 hit songs have different audio-feature patterns from the broader Spotify track sample. Our project will not claim that audio features cause popularity. Instead, we will study the association between audio features and popularity score, and we will discuss the limits of this relationship.

1. <a name="cite_note-1"></a> [^](#cite_ref-1) Spotify for Developers. "Get Track." Spotify Web API Reference. https://developer.spotify.com/documentation/web-api/reference/get-track  
2. <a name="cite_note-2"></a> [^](#cite_ref-2) Pandya, M. "Spotify Tracks Dataset." Kaggle. https://www.kaggle.com/datasets/maharshipandya/-spotify-tracks-dataset  
3. <a name="cite_note-3"></a> [^](#cite_ref-3) Pandya, M. "maharshipandya/spotify-tracks-dataset." Hugging Face Datasets. https://huggingface.co/datasets/maharshipandya/spotify-tracks-dataset  
4. <a name="cite_note-4"></a> [^](#cite_ref-4) Spotify for Developers. "Search for Item." Spotify Web API Reference. https://developer.spotify.com/documentation/web-api/reference/search  
5. <a name="cite_note-5"></a> [^](#cite_ref-5) Nelgiriyewithana, N. "Most Streamed Spotify Songs 2024." Kaggle. https://www.kaggle.com/datasets/nelgiriyewithana/most-streamed-spotify-songs-2024
6. <a name="cite_note-6"></a> [^](#cite_ref-6) Saragih, H. S. "Predicting Song Popularity Based on Spotify's Audio Features." *Journal of Management Analytics*, 2023. https://www.tandfonline.com/doi/full/10.1080/23270012.2023.2239824
7. <a name="cite_note-7"></a> [^](#cite_ref-7) Pachet, F., and Roy, P. "Hit Song Science Is Not Yet a Science." *Proceedings of the 9th International Society for Music Information Retrieval Conference*, 2008. https://www.cs.swarthmore.edu/~turnbull/cs97/f08/paper/pachet08.pdf
8. <a name="cite_note-8"></a> [^](#cite_ref-8) Duman, D., Neto, P., Mavrolampados, A., Toiviainen, P., and Luck, G. "Music We Move To: Spotify Audio Features and Reasons for Listening." *PLOS ONE*, 2022. https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0275228

## Hypothesis


We hypothesize that Spotify audio features are associated with Spotify popularity score, but this relationship will be moderate rather than strong. Specifically, we expect danceability, energy, and loudness to show positive relationships with popularity score. In other words, songs that are more danceable, energetic, and louder may have higher popularity scores on average.

This hypothesis is based on the idea that these audio features may make songs more suitable for repeated listening, playlists, social settings, and mainstream streaming environments. Danceable songs may be easier to play in social or casual listening contexts. Energetic songs may feel more engaging to listeners. Louder songs may also stand out more in streaming and playlist settings. For these reasons, we expect these features to have some connection with popularity.

As a supplementary comparison, we also expect matched songs from the Most Streamed Spotify Songs 2024 dataset to have higher average danceability, energy, and loudness than the broader Spotify track sample. However, this comparison depends on whether enough songs can be matched successfully between the two datasets.

We do not expect audio features to fully explain popularity. Popularity is also affected by non-audio factors such as artist fame, playlist placement, promotion, TikTok trends, release timing, and platform recommendations. Since these factors are not fully captured in our audio-feature dataset, our expectation is that audio features will explain only part of the variation in popularity score.

## Data

### Data overview

- **Dataset #1: Spotify Tracks Dataset**
  - Dataset Name: Spotify Tracks Dataset
  - Link: https://www.kaggle.com/datasets/maharshipandya/-spotify-tracks-dataset
  - Number of observations: 114,000 rows in the raw dataset; 89,740 rows after removing the row with missing track information and deduplicating by `track_id`
  - Number of variables: 21 in the raw dataset; 20 after dropping the unnecessary index column
  - Most relevant variables: `popularity` is the main outcome variable in our project. It is a Spotify popularity score from 0 to 100. The main predictor variables are the Spotify audio features, including `danceability`, `energy`, `valence`, `acousticness`, `speechiness`, `instrumentalness`, and `liveness`, which are mostly on a 0–1 scale. Other useful variables include `loudness`, which is measured in decibels, `tempo`, which is measured in beats per minute, and `duration_ms`, which is measured in milliseconds. The dataset also includes `track_genre`, which can help us understand the data during EDA.
  - Shortcomings: One issue with this dataset is that it is not a natural sample of all Spotify music. It has 114 genre labels with exactly 1,000 tracks in each genre, so the genre distribution is artificially balanced. This probably does not match the real distribution of tracks on Spotify. Another issue is that many tracks appear more than once under the same `track_id`, often because they are listed under multiple genres. These duplicate rows need to be removed before our main analysis. Also, Spotify popularity score reflects recent play activity, so it is not the same as total lifetime streams. Finally, genre will be useful for understanding the dataset, but it is not the main thing we are trying to predict.

- **Dataset #2: Most Streamed Spotify Songs 2024**
  - Dataset Name: Most Streamed Spotify Songs 2024
  - Link: https://www.kaggle.com/datasets/nelgiriyewithana/most-streamed-spotify-songs-2024
  - Number of observations: 4,600 rows
  - Number of variables: 29
  - Most relevant variables: `Track`, `Artist`, `Spotify Streams`, `Spotify Popularity`, `YouTube Views`, and `TikTok Views`. `Spotify Streams` gives the total Spotify stream count, while `Spotify Popularity` is a 0–100 Spotify popularity score. The YouTube and TikTok variables are also useful because they give us some information about a song’s popularity outside Spotify.
  - Shortcomings: This dataset is useful for finding recent hit songs, but it does not include the full set of Spotify audio features that we need for our main regression analysis. Because of that, it cannot replace the Spotify Tracks Dataset. It also has a lot of missing values in some platform-specific columns. For example, `TIDAL Popularity` is completely missing, `Soundcloud Streams` is missing for about 72% of the rows, and several other platform columns have around 20–46% missing values. Some numeric columns, such as `Spotify Streams`, are stored as strings with commas, so they need to be converted into numeric values before analysis. This dataset is also biased toward mainstream 2024 hit songs, so it does not represent the broader range of Spotify tracks.

**Combining the datasets:** We plan to match songs from the 2024 dataset to the Spotify Tracks Dataset by cleaning and joining on track name and artist. The 2024 dataset does not include the full audio features, so after matching, the matched 2024 songs will get their audio features from the Spotify Tracks Dataset. Our current matching process gives 1,311 matched rows after merging. However, this does not necessarily mean 1,311 unique hit songs, because duplicate track names, repeated tracks, different versions, or genre duplicates may create multiple matched rows. We will use this matched subset only as a supplementary comparison, not as the main analysis dataset. The main analysis still focuses on the relationship between Spotify audio features and popularity score in the cleaned Spotify Tracks sample.

In [1]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [2]:
import os

# Data files are stored in data/00-raw/ in this repository.
# No download is needed if the files are already present.
required_files = [
    'data/00-raw/spotify_tracks.csv',
    'data/00-raw/spotify_2024.csv',
]

for path in required_files:
    exists = os.path.exists(path)
    size_mb = os.path.getsize(path) / 1e6 if exists else 0
    status = f'OK  ({size_mb:.1f} MB)' if exists else 'MISSING'
    print(f'{status}  {path}')


OK  (20.1 MB)  data/00-raw/spotify_tracks.csv
OK  (1.1 MB)  data/00-raw/spotify_2024.csv


### Dataset #1: Spotify Tracks Dataset

This dataset contains 114,000 tracks collected from Spotify-related track data and organized across 114 music genres with 1,000 tracks per genre. Each row represents one track-genre pairing, meaning the same track can appear more than once if it is tagged under multiple genres. After dropping the row with missing track information and removing duplicate `track_id` entries, the cleaned dataset has 89,740 unique tracks.

**Key variables and their meanings:**

- `popularity`: Integer from 0 to 100. Spotify's internal popularity score, calculated based mostly on total play count and how recent those plays are. A score of 100 is the most popular. This is our main outcome variable.
- `danceability`: Float from 0.0 to 1.0. Describes how suitable a track is for dancing, based on tempo, rhythm stability, beat strength, and regularity. Higher values indicate more danceable tracks.
- `energy`: Float from 0.0 to 1.0. Measures perceived intensity and activity. High-energy tracks feel fast, loud, and noisy; low-energy tracks feel calm and quiet.
- `loudness`: Overall loudness of the track in decibels (dB). Less negative values usually mean louder tracks.
- `valence`: Float from 0.0 to 1.0. Describes musical positiveness. High valence means the track sounds happier or more cheerful, while low valence means it sounds sadder, angrier, or more tense.
- `tempo`: Estimated beats per minute (BPM). Most normal songs fall roughly between 50 and 220 BPM, but we will check unusual values during cleaning.
- `acousticness`: Float from 0.0 to 1.0. Confidence that the track is acoustic. Values close to 1.0 mean the track is likely acoustic.
- `speechiness`: Float from 0.0 to 1.0. Detects spoken words. Values above 0.66 suggest mostly spoken word content; values below 0.33 suggest music with little or no speech.
- `instrumentalness`: Float from 0.0 to 1.0. Predicts whether the track has no vocals. Values above 0.5 suggest the track is likely instrumental.
- `liveness`: Float from 0.0 to 1.0. Detects whether a live audience is present. Values above 0.8 strongly suggest a live recording.
- `duration_ms`: Track duration in milliseconds. Most tracks fall between 60,000 ms (1 minute) and 600,000 ms (10 minutes), but very short or very long tracks will be checked during cleaning.
- `explicit`: Boolean. True if the track contains explicit lyrics.
- `key`: Integer from 0 to 11. Estimated overall key of the track (0 = C, 1 = C#/Db, …, 11 = B).
- `mode`: 1 = major, 0 = minor.
- `time_signature`: Estimated number of beats per measure. In this dataset, common values include 3, 4, and 5, while unusual values such as 0 or 1 need to be checked.
- `track_genre`: Categorical. One of 114 genre labels assigned to the track.

**Data concerns:** The genre-balanced design, with 1,000 tracks per genre, means results may not generalize to the actual genre distribution on Spotify. Popularity scores are also skewed toward lower values for many tracks, suggesting the dataset includes many low-visibility songs. One row has missing values in `artists`, `album_name`, and `track_name`, and it is dropped during cleaning. Another important concern is duplication: many tracks appear under multiple genres with the same `track_id`, so we remove duplicate `track_id` entries before the main analysis.

In [3]:
import pandas as pd
import numpy as np
import os

# --- Load raw data ---
tracks_raw = pd.read_csv('data/00-raw/spotify_tracks.csv')
print('Raw shape:', tracks_raw.shape)
print('Columns:', list(tracks_raw.columns))

# --- Missing values ---
print('\nMissing values per column:')
print(tracks_raw.isnull().sum()[tracks_raw.isnull().sum() > 0])

# --- Check duplicate track_ids ---
print(f'\nDuplicate track_id rows: {tracks_raw["track_id"].duplicated().sum()}')
print('(Same track can appear under multiple genres, so duplicate track_id rows need to be removed.)')

# --- Clean ---
tracks = tracks_raw.drop(columns=['Unnamed: 0'])

# Drop the row with missing artist/album/track name
tracks = tracks.dropna(subset=['artists', 'album_name', 'track_name'])

# Keep only one row per track_id
# The first occurrence preserves one genre label for each unique track
tracks = tracks.drop_duplicates(subset=['track_id']).reset_index(drop=True)

print('\nCleaned shape:', tracks.shape)
print('Missing values after cleaning:', tracks.isnull().sum().sum())

# --- Key variable summaries ---
print('\nPopularity summary:')
print(tracks['popularity'].describe().round(2))

features_01 = [
    'danceability', 'energy', 'valence', 'acousticness',
    'speechiness', 'instrumentalness', 'liveness'
]

features_other = ['loudness', 'tempo', 'duration_ms']

print('\n0-1 audio feature ranges:')
print(tracks[features_01].agg(['min', 'max', 'mean']).round(3))

print('\nOther audio-related feature ranges:')
print(tracks[features_other].agg(['min', 'max', 'mean']).round(3))

# --- Check expected ranges / suspicious values ---
print('\nRange checks:')
print('Popularity outside 0-100:', ((tracks['popularity'] < 0) | (tracks['popularity'] > 100)).sum())

for col in features_01:
    invalid_count = ((tracks[col] < 0) | (tracks[col] > 1)).sum()
    print(f'{col} outside 0-1:', invalid_count)

print('Duration <= 0:', (tracks['duration_ms'] <= 0).sum())
print('Tempo <= 0:', (tracks['tempo'] <= 0).sum())
print('Loudness > 0:', (tracks['loudness'] > 0).sum())

# --- Save cleaned dataset ---
os.makedirs('data/02-processed', exist_ok=True)
tracks.to_csv('data/02-processed/spotify_tracks_clean.csv', index=False)

print('\nSaved: data/02-processed/spotify_tracks_clean.csv')
print('Final shape:', tracks.shape)

Raw shape: (114000, 21)
Columns: ['Unnamed: 0', 'track_id', 'artists', 'album_name', 'track_name', 'popularity', 'duration_ms', 'explicit', 'danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness', 'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'time_signature', 'track_genre']

Missing values per column:
artists       1
album_name    1
track_name    1
dtype: int64

Duplicate track_id rows: 24259
(Same track can appear under multiple genres, so duplicate track_id rows need to be removed.)

Cleaned shape: (89740, 20)
Missing values after cleaning: 0

Popularity summary:
count    89740.00
mean        33.20
std         20.58
min          0.00
25%         19.00
50%         33.00
75%         49.00
max        100.00
Name: popularity, dtype: float64

0-1 audio feature ranges:
      danceability  energy  valence  acousticness  speechiness  \
min          0.000   0.000    0.000         0.000        0.000   
max          0.985   1.000    0.995         0.996        0

After cleaning, the Spotify Tracks Dataset has 89,740 rows and 20 columns. We removed the extra index column, dropped the one row with missing track information, and removed duplicate `track_id` rows so each track only appears once in the main dataset.

The main outcome variable, `popularity`, is valid from 0 to 100. Its mean is about 33.20 and the median is 33, so many songs in this dataset have relatively low popularity scores. The 0–1 audio features, such as `danceability`, `energy`, `valence`, `acousticness`, `speechiness`, `instrumentalness`, and `liveness`, are all within their expected ranges.

We also found 157 rows with `tempo` equal to 0 and 68 rows with positive `loudness`. We keep these rows for now because they are a small part of the dataset and may just be unusual Spotify estimates rather than clear errors. We will be careful when using these variables later.

### Dataset #2: Most Streamed Spotify Songs 2024 + Dataset Merging

This dataset contains 4,600 highly streamed songs from 2024, along with cross-platform performance metrics from Spotify, YouTube, TikTok, Apple Music, Deezer, Amazon, Pandora, SiriusXM, Soundcloud, and Shazam. In our project, this dataset is not the main audio-feature dataset. Instead, we use it to identify recent hit songs and then compare those matched hit songs with the broader Spotify Tracks sample.

**Key variables and their meanings:**

- `Track`, `Artist`, `Album Name`, `Release Date`: Track identifiers. `Track` and `Artist` are especially important because we use them to match songs to the Spotify Tracks Dataset.
- `ISRC`: International Standard Recording Code. This is a unique global identifier for recordings, but we do not use it as the main merge key because the Spotify Tracks Dataset does not include the same ISRC column.
- `All Time Rank`: Ranking variable provided by the dataset for the listed songs.
- `Track Score`: A dataset-provided score related to cross-platform performance. Since the exact formula is not fully explained, we treat it as a descriptive variable instead of our main outcome variable.
- `Spotify Streams`: Total Spotify stream count. In the raw file, this column is stored as a comma-formatted string, so it needs to be converted into numeric form during cleaning.
- `Spotify Popularity`: Spotify popularity score from 0 to 100. This is useful for describing the 2024 hit songs, but it has missing values.
- `YouTube Views`, `TikTok Views`: View or play counts on those platforms. These variables help describe cross-platform popularity, but they also have missing values.
- `Explicit Track`: A 0 or 1 flag indicating whether the track contains explicit content.

**Missing data:** This dataset has many missing values in platform-specific columns. We do not drop every row with missing values because many of those columns are not needed for our main analysis or for matching songs. Missing values are kept as `NaN` instead of being filled with 0, because missing data means the value was not recorded, not that the true value is zero. For example, a missing value in `TikTok Views` does not necessarily mean the song had zero TikTok views.

We drop `TIDAL Popularity` because it is completely missing. We also drop rows that are missing `Track` or `Artist`, because those rows cannot be matched reliably to the Spotify Tracks Dataset. Other platform metrics, such as `Soundcloud Streams`, `Pandora Streams`, and `SiriusXM Spins`, are kept only as optional descriptive variables and are not used as main predictors. Numeric columns such as `Spotify Streams`, `Spotify Popularity`, `YouTube Views`, and `TikTok Views` are converted from comma-formatted strings into numeric values.

**Dataset concerns:** This dataset is biased toward globally popular and mainstream songs from 2024. It does not represent all Spotify songs, and it also does not represent less-known songs, older songs, or independent artists very well. Another important limitation is that this dataset does not include the full Spotify audio features used in our main regression analysis. Because of that, it cannot replace the Spotify Tracks Dataset. Its role is only to help us identify recent hit songs for a supplementary comparison.

**Merging strategy:** We merge this dataset with the cleaned Spotify Tracks Dataset using a normalized inner join on `track_name` and `artist`, where both fields are lowercased and stripped of extra spaces. Songs that cannot be matched are excluded from the merged subset. Since the 2024 dataset does not contain the full audio-feature profile, matched songs get their audio features from the cleaned Spotify Tracks Dataset.

The current merge produces 1,311 matched rows. However, this does not mean 1,311 unique hit songs, because duplicate entries, repeated tracks, different versions, or slightly different track records can create multiple matched rows. We use this merged dataset only as a supplementary comparison. The main analysis still focuses on the relationship between Spotify audio features and popularity score in the cleaned Spotify Tracks Dataset. The merged dataset is saved as `data/02-processed/spotify_merged_clean.csv`.

In [4]:
import pandas as pd
import numpy as np
import os

# --- Load raw 2024 hits data ---
hits_raw = pd.read_csv('data/00-raw/spotify_2024.csv', encoding='latin1')
print('Raw shape:', hits_raw.shape)

print('\nMissing values per column:')
missing = hits_raw.isnull().sum()
print(missing[missing > 0].sort_values(ascending=False))

# --- Select relevant columns ---
# We do not keep TIDAL Popularity because it is 100% missing.
cols_keep = [
    'Track', 'Artist', 'Album Name', 'Release Date', 'ISRC',
    'All Time Rank', 'Track Score',
    'Spotify Streams', 'Spotify Popularity',
    'YouTube Views', 'TikTok Views', 'Explicit Track'
]

hits = hits_raw[cols_keep].copy()

# --- Drop rows that cannot be matched ---
# Track and Artist are needed for matching with the Spotify Tracks Dataset.
hits = hits.dropna(subset=['Track', 'Artist']).reset_index(drop=True)

# --- Convert comma-formatted numeric columns to numeric ---
# Missing values stay as NaN. We do not fill them with 0.
numeric_cols = [
    'All Time Rank', 'Track Score',
    'Spotify Streams', 'Spotify Popularity',
    'YouTube Views', 'TikTok Views',
    'Explicit Track'
]

for col in numeric_cols:
    hits[col] = (
        hits[col]
        .astype(str)
        .str.replace(',', '', regex=False)
    )
    hits[col] = pd.to_numeric(hits[col], errors='coerce')

print('\n2024 hits shape after cleaning:', hits.shape)

print('\nMissing values after cleaning:')
missing_clean = hits.isnull().sum()
print(missing_clean[missing_clean > 0].sort_values(ascending=False))

# --- Save cleaned 2024 dataset ---
os.makedirs('data/02-processed', exist_ok=True)
hits.to_csv('data/02-processed/spotify_2024_clean.csv', index=False)
print('\nSaved: data/02-processed/spotify_2024_clean.csv')

# --- Load cleaned tracks dataset ---
tracks = pd.read_csv('data/02-processed/spotify_tracks_clean.csv')

# --- Normalize keys for matching ---
# Simple normalization: lowercase and remove extra spaces.
hits['_track_key'] = hits['Track'].str.lower().str.strip()
hits['_artist_key'] = hits['Artist'].str.lower().str.strip()

tracks['_track_key'] = tracks['track_name'].str.lower().str.strip()
tracks['_artist_key'] = tracks['artists'].str.lower().str.strip()

# --- Inner join on normalized track + artist ---
merged = pd.merge(
    hits,
    tracks,
    on=['_track_key', '_artist_key'],
    how='inner'
)

# Drop temporary matching keys
merged = merged.drop(columns=['_track_key', '_artist_key'])

print('\nMerged shape:', merged.shape)
print('Unique matched Track + Artist pairs:', merged[['Track', 'Artist']].drop_duplicates().shape[0])
print('Unique matched track_id values:', merged['track_id'].nunique())

print('\nMissing values in merged dataset:')
missing_m = merged.isnull().sum()
print(missing_m[missing_m > 0].sort_values(ascending=False))

# --- Save merged dataset ---
merged.to_csv('data/02-processed/spotify_merged_clean.csv', index=False)
print('\nSaved: data/02-processed/spotify_merged_clean.csv')

Raw shape: (4600, 29)

Missing values per column:
TIDAL Popularity              4600
Soundcloud Streams            3333
SiriusXM Spins                2123
Pandora Track Stations        1268
TikTok Posts                  1173
Pandora Streams               1106
Amazon Playlist Count         1055
YouTube Playlist Reach        1009
TikTok Views                   981
TikTok Likes                   980
Deezer Playlist Reach          928
Deezer Playlist Count          921
Spotify Popularity             804
Shazam Counts                  577
Apple Music Playlist Count     561
AirPlay Spins                  498
YouTube Likes                  315
YouTube Views                  308
Spotify Streams                113
Spotify Playlist Reach          72
Spotify Playlist Count          70
Artist                           5
dtype: int64

2024 hits shape after cleaning: (4595, 12)

Missing values after cleaning:
TikTok Views          976
Spotify Popularity    799
YouTube Views         303
Spotify Strea

## Ethics

[![Deon badge](https://img.shields.io/badge/ethics%20checklist-deon-brightgreen.svg?style=popout-square)](http://deon.drivendata.org/)

### A. Data Collection
 - [ ] **A.1 Informed consent**: If there are human subjects, have they given informed consent, where subjects affirmatively opt-in and have a clear understanding of the data uses to which they consent?

 - [X] **A.2 Collection bias**: Have we considered sources of bias that could be introduced during data collection and survey design and taken steps to mitigate those?

> Our datasets only represent Spotify-related data and may not reflect all music platforms or all listeners worldwide. The 2024 most-streamed dataset is also biased toward already successful mainstream songs, which means less popular or non-English music is underrepresented in our analysis.

 - [X] **A.3 Limit PII exposure**: Have we considered ways to minimize exposure of personally identifiable information (PII) for example through anonymization or not collecting information that isn't relevant for analysis?

> Our datasets use public track-level data and do not include any private listener information such as names, locations, emails, or personal listening histories. No PII is present in either dataset.

 - [ ] **A.4 Downstream bias mitigation**: Have we considered ways to enable testing downstream results for biased outcomes (e.g., collecting data on protected group status like race or gender)?

### B. Data Storage
 - [ ] **B.1 Data security**: Do we have a plan to protect and secure data (e.g., encryption at rest and in transit, access controls on internal users and third parties, access logs, and up-to-date software)?
 - [ ] **B.2 Right to be forgotten**: Do we have a mechanism through which an individual can request their personal information be removed?
 - [ ] **B.3 Data retention plan**: Is there a schedule or plan to delete the data after it is no longer needed?

### C. Analysis
 - [ ] **C.1 Missing perspectives**: Have we sought to address blindspots in the analysis through engagement with relevant stakeholders (e.g., checking assumptions and discussing implications with affected communities and subject matter experts)?
 - [X] **C.2 Dataset bias**: Have we examined the data for possible sources of bias and taken steps to mitigate or address these biases (e.g., stereotype perpetuation, confirmation bias, imbalanced classes, or omitted confounding variables)?

> Some genres may have many more tracks than others in our dataset, which could skew the regression model toward predicting popularity for overrepresented genres. We will check genre distribution during EDA and note any imbalance in our analysis.

 - [X] **C.3 Honest representation**: Are our visualizations, summary statistics, and reports designed to honestly represent the underlying data?

> We will not claim that audio features cause popularity. Our analysis can only show prediction or association. All visualizations and summary statistics will be presented with appropriate labels and without cherry-picking results to support a preferred conclusion.

 - [ ] **C.4 Privacy in analysis**: Have we ensured that data with PII are not used or displayed unless necessary for the analysis?
 - [X] **C.5 Auditability**: Is the process of generating the analysis well documented and reproducible if we discover issues in the future?

> All data cleaning, merging, visualization, and modeling steps will be kept in the notebook with clear comments so the full analysis can be checked and reproduced by others.

### D. Modeling
 - [ ] **D.1 Proxy discrimination**: Have we ensured that the model does not rely on variables or proxies for variables that are unfairly discriminatory?
 - [ ] **D.2 Fairness across groups**: Have we tested model results for fairness with respect to different affected groups (e.g., tested for disparate error rates)?
 - [X] **D.3 Metric selection**: Have we considered the effects of optimizing for our defined metrics and considered additional metrics?

> Spotify popularity score is our main outcome variable, but it is not the same as total lifetime streams since it weighs recent plays more heavily. We will note this limitation and consider using Spotify streams from the 2024 dataset as a secondary metric where possible.

 - [X] **D.4 Explainability**: Can we explain in understandable terms a decision the model made in cases where a justification is needed?

> We plan to use regression-based methods, so we can explain which audio features are associated with higher or lower predicted popularity scores in plain terms. We will avoid using black-box models without explanation.

 - [X] **D.5 Communicate limitations**: Have we communicated the shortcomings, limitations, and biases of the model to relevant stakeholders in ways that can be generally understood?

> We will clearly state that popularity is also affected by artist fame, playlist placement, promotion, social media trends, and release timing, and that these factors are not captured in our model. We will also note that matching songs across the two datasets by track name and artist may be imperfect.

### E. Deployment
 - [ ] **E.1 Monitoring and evaluation**: Do we have a clear plan to monitor the model and its impacts after it is deployed (e.g., performance monitoring, regular audit of sample predictions, human review of high-stakes decisions, reviewing downstream impacts of errors or low-confidence decisions, testing for concept drift)?
 - [ ] **E.2 Redress**: Have we discussed with our organization a plan for response if users are harmed by the results (e.g., how does the data science team evaluate these cases and update analysis and models to prevent future harm)?
 - [ ] **E.3 Roll back**: Is there a way to turn off or roll back the model in production if necessary?
 - [X] **E.4 Unintended use**: Have we taken steps to identify and prevent unintended uses and abuse of the model and do we have a plan to monitor these once the model is deployed?

> Our project should not be treated as a guaranteed formula for making a popular song. We will present our results as patterns observed in these specific datasets, not as universal rules applicable to all music or all platforms.

## Team Expectations 

Our team will mainly communicate through iMessage and meet when needed, usually once a week or before major deadlines. We expect everyone to reply to important messages within about 24 hours and finish their assigned work on time.

We will divide work based on each person's strengths, but everyone should contribute to research, writing, data cleaning, analysis, and editing at some point. If someone is confused or falling behind, they should tell the group early so we can help or redistribute the work.

For decisions, we will discuss as a group and choose the option that most people agree with. We want our communication to be honest but respectful, especially when giving feedback or disagreeing with an idea.

## Project Timeline Proposal

| Date | Task |
|---|---|
| Apr 29 | Submit project proposal. Finalize research question, hypothesis, background, data sources, ethics, team expectations, and timeline. |
| May 1–May 6 | Download both Spotify datasets, inspect columns, check missing values, and decide which variables to keep. |
| **Sun May 3** | **Team meeting: review dataset structure, assign cleaning tasks for the week.** |
| May 7–May 13 | Clean the Spotify Tracks Dataset and prepare the Data Checkpoint. Create basic summaries of popularity score, genres, and audio features. |
| **Sun May 10** | **Team meeting: review data cleaning progress, finalize variables for Data Checkpoint.** |
| May 14–May 20 | Start exploratory data analysis. Make visualizations for popularity, audio features, and genre differences. |
| **Sun May 17** | **Team meeting: review EDA visualizations, coordinate next steps for analysis.** |
| May 21–May 27 | Finish EDA Checkpoint. Explore correlations between audio features and popularity, and begin testing dataset matching by track name and artist. |
| **Sun May 24** | **Team meeting: review correlation results, finalize EDA Checkpoint submission.** |
| May 28–Jun 3 | Build regression models to predict Spotify popularity score and compare audio features between 2024 most-streamed songs and the broader Spotify catalog if matching works. |
| **Sun May 31** | **Team meeting: review model results, discuss findings and interpret coefficients.** |
| Jun 4–Jun 10 | Finish final report and video. Write results, discussion, limitations, ethics reflection, and final edits before submission. |
| **Sun Jun 7** | **Team meeting: final review of report and video before submission.** |